File: src/main/java/com/github/sorusclient/client/ui/framework/constraint/Relative.java
package com.github.sorusclient.client.ui.framework.constraint;

import com.github.sorusclient.client.ui.framework.Component;
import com.github.sorusclient.client.util.Color;

public class Relative implements Constraint {

    private final double percent;
    private final boolean otherSide;

    public Relative(double percent) {
        this(percent, false);
    }

    public Relative(double percent, boolean otherSide) {
        this.percent = percent;
        this.otherSide = otherSide;
    }

    @Override
    public double getXValue(Component.Runtime componentRuntime) {
        return componentRuntime.getParent().getWidth() * this.percent;
    }

    @Override
    public double getYValue(Component.Runtime componentRuntime) {
        return componentRuntime.getParent().getHeight() * this.percent;
    }

    @Override
    public double getWidthValue(Component.Runtime componentRuntime) {
        double size = this.otherSide ? componentRuntime.getParent().getHeight() : componentRuntime.getParent().getWidth();
        return size * this.percent;
    }

    @Override
    public double getHeightValue(Component.Runtime componentRuntime) {
        double size = this.otherSide ? componentRuntime.getParent().getWidth() : componentRuntime.getParent().getHeight();
        return size * this.percent;
    }

    @Override
    public double getPaddingValue(Component.Runtime componentRuntime) {
        double size = this.otherSide ? componentRuntime.getParent().getHeight() : componentRuntime.getParent().getWidth();
        return size * this.percent;
    }

    @Override
    public Color getColorValue(Component.Runtime componentRuntime) {
        return null;
    }

    @Override
    public String getStringValue(Component.Runtime componentRuntime) {
        return null;
    }

    @Override
    public double getCornerRadiusValue(Component.Runtime componentRuntime) {
        return this.getPaddingValue(componentRuntime);
    }

}

File: src/main/kotlin/com/github/sorusclient/client/ui/framework/constraint/Relative.kt
package com.github.sorusclient.client.ui.framework.constraint

import com.github.sorusclient.client.ui.framework.Component
import com.github.sorusclient.client.util.Color

class Relative @JvmOverloads constructor(private val percent: Double, private val otherSide: Boolean = false) :
    Constraint {
    override fun getXValue(componentRuntime: Component.Runtime): Double {
        return componentRuntime.getParent()!!.width * percent
    }

    override fun getYValue(componentRuntime: Component.Runtime): Double {
        return componentRuntime.getParent()!!.height * percent
    }

    override fun getWidthValue(componentRuntime: Component.Runtime): Double {
        val size = if (otherSide) componentRuntime.getParent()!!
            .height else componentRuntime.getParent()!!.width
        return size * percent
    }

    override fun getHeightValue(componentRuntime: Component.Runtime): Double {
        val size = if (otherSide) componentRuntime.getParent()!!
            .width else componentRuntime.getParent()!!.height
        return size * percent
    }

    override fun getPaddingValue(componentRuntime: Component.Runtime): Double {
        val size = if (otherSide) componentRuntime.getParent()!!
            .height else componentRuntime.getParent()!!.width
        return size * percent
    }

    override fun getColorValue(componentRuntime: Component.Runtime): Color {
        return null!!
    }

    override fun getStringValue(componentRuntime: Component.Runtime): String {
        return null!!
    }

    override fun getCornerRadiusValue(componentRuntime: Component.Runtime): Double {
        return getPaddingValue(componentRuntime)
    }
}